rfd/0083-license-expiration-warnings.md

```diff
@@ -0,0 +1,153 @@
+CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592
+
+Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.
+```
+
+
+CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.
+
+### Web UI UX
+
+Web UI will display license expiration banner on top of the page with the following rules:
+Show warning (yellow) to users N days prior to expiration, 1 day snooze
+Show error (red) to all users after expiration, no dismiss or snoozing available.
+Snoozing allows users to disable the warning in the web UI from being displayed for N days.
+
+## Implementation details
+
+License warnings can piggyback on the cluster alert endpoint `ServerWithRoles.GetClusterAlerts` with the responses being
+
+```
+{
+  "alerts": [
+    {
+      "kind": "cluster_alert",
+      "version": "v1",
+      "metadata": {
+        "name": "123e4567-e89b-12d3-a456-426614174000.",
+        "labels": {
+          "teleport.internal/alert-on-login": "yes",
+          "teleport.internal/alert-permit-all": "yes"
+        },
+        "expires": "2022-08-31T17:26:05.728149Z"
+      },
+      "spec": {
+        "severity": 5,
+        "message": "Your Teleport Enterprise Edition license will expire in 10 days. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.",
+        "created": "2022-08-30T17:26:05.728149Z"
+      }
+    }
+  ]
+}
+```
+
+```
+{
+  "alerts": [
+    {
+      "kind": "cluster_alert",
+      "version": "v1",
+      "metadata": {
+        "name": "123e4567-e89b-12d3-a456-426614174000",
+        "labels": {
+          "teleport.internal/alert-on-login": "yes",
+          "teleport.internal/alert-permit-all": "yes",
+          "teleport.internal/license-warning-expired": "yes",
+        },
+        "expires": "2022-08-31T17:26:05.728149Z"
+      },
+      "spec": {
+        "severity": 5,
+        "message": "Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.",
+        "created": "2022-08-30T17:26:05.728149Z"
+      }
+    }
+  ]
+}
+```
+
+
+On startup and every 1 hour afterwards the auth server will check the license and generate a license warning if applicable, or clear license warning alerts if they are no longer needed.
+
+These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this.
+
+## Security considerations
+If your license expired, you can’t disable or snooze the warning.
```

Comment:
We will need to modify the `alert` spec to include a `dismissible` boolean flag. 

We can cheat it for now to say something like `high` severity alerts are not dismissible but I think adding a flag would be more resilient.

Reply:
Could the non dismissible alerts be determined by a label?

Reply:
Discussed this over slack with hatched, going to go with the boolean flag